frontend/src/app/api/attachment-download/[opportunityId]/route.ts

```diff
@@ -0,0 +1,81 @@
+import { randomUUID } from "crypto";
+import * as zip from "@zip.js/zip.js";
+import { ApiRequestError, readError } from "src/errors";
+import { getOpportunityDetails } from "src/services/fetch/fetchers/opportunityFetcher";
+
+import { NextRequest, NextResponse } from "next/server";
+
+export async function GET(
+  request: NextRequest,
+  { params }: { params: Promise<{ opportunityId: string }> },
+): Promise<Response> {
+  const { opportunityId } = await params;
+  try {
+    const response = await getOpportunityDetails(opportunityId);
+
+    if (!response || response.status_code !== 200) {
+      throw new ApiRequestError(
+        `Error fetching Opportunity: ${response.message}`,
+        "APIRequestError",
+        response.status_code,
+      );
+    }
+
+    if (response.data.attachments?.length === 0) {
+      return Response.json(
+        {
+          message: `No files for Opportunity`,
+          opportunityId,
+        },
+        { status: 404 },
+      );
+    }
+    const zipWriter = new zip.ZipWriterStream();
+
+    const closeWriter = async () => {
+      await zipWriter.close();
+    };
+
+    Promise.allSettled(
```

Comment:
What about a refactor that makes use of awaits here to clean up the nested promises and makes this a little easier to read. Something like ... oops this got kind of involved
```suggestion
    const addAttachmentToZip = (attachment) =>
      new zip.HttpReader(attachment.download_path).readable.pipeTo(
        zipWriter.writable(attachment.file_name),
      );

    await Promise.allSettled(
      response.data.attachments.map((attachment) => {
        return addAttachmentToZip(attachment)
          .catch((e: Error) => {
            const uuid = randomUUID();
            console.error(
              "Error fetching file",
              attachment.download_path,
              uuid,
            );
            throw new Error(`${e.message}, ${uuid}`);
          })
          .finally(() => {
            closeWriter();
          });
      }),
    );
```

Reply:
so my code above doesn't work, while yours basically does which is a big stumper. I can't seem to get the stream functions to work as promises once you actually wait for them, but they work when they are not awaited. For some reason only one promise ever gets resolved once you start to await.

So I abandoned that and tried a solution without streams that seems to work fine (if not better - for some reason the file size of the zip is slightly smaller in my testing). How's this?

```
    const blobWriter = new zip.BlobWriter();
    const zipWriter = new zip.ZipWriter(blobWriter);
    const addPromises = response.data.attachments?.map((attachment) => {
      return zipWriter.add(
        attachment.file_name,
        new zip.HttpReader(attachment.download_path),
      );
    });

    await Promise.all([...addPromises, zipWriter.close()]);

    const blobData = await blobWriter.getData();
    return new NextResponse(blobData, {
```